[docs] Add guide on how to use MUI Base with Tailwind CSS (#33100)
mui/material-ui: docs/data/base/guides/interoperability-with-tailwind-css/interoperability-with-tailwind-css.md

> # Interoperability with Tailwind CSS

Reviewer: ```suggestion
# Using Tailwind CSS
```
I figure that technically "_interoperability_" might be what happens (two (not necessarily designed to work together) libraries working together?) but this word is terribly confusing and complex to me 😅 Guess we could go with something simpler, what y'all think?

Author: I will leave the decision to @samuelsycamore :D I will update the files + pages name together based on what we decide to go with :)

Reviewer 2: +1 for replacing "Interoperability" here and everywhere else in our docs! In this case I think I would lean toward "Working with" but that's very compared to "Using."

```suggestion
# Working with Tailwind CSS
```

> <p class="description">Learn the basics of how to use Tailwind CSS with MUI Base components.</p>

Reviewer: ```suggestion
<p class="description">Learn how to use Tailwind CSS with MUI Base components.</p>
```
Removing "the basics" because it kinda implies there are further details I'd want to check and this is sort of an introduction to the subject.

## Getting started

The goal of this guide is to teach you how to use Tailwind CSS with MUI Base.

> The guide assumes you have experience with Tailwind CSS already, or you are at least familiar with it.

Reviewer: ```suggestion
If you don't have experience with Tailwind CSS, we recommend you study that first before diving into this guide.
```
A different way to express the same idea?!

> We are not going to cover how to build the UI using Tailwind CSS, but rather how to MUI Base on top of a view already biult with Tailwind CSS, in order to make it interactive and accessible.

Reviewer: ```suggestion
We're not going to cover all the steps to build the UI we'll use as an example but rather how to insert MUI Base components, making it interactive and accessible, and style them with Tailwind CSS.
```

Trough out the guide, we'll build an interactive player.
You can see the final look of it on the picture below.

<img src="/static/images/base-with-tailwind-css/player-final.png" alt="Example UI using Tailwind CSS - player" width="638" />

:::info

> The player is took as an inspiration from the [Tailwind CSS](https://tailwindcss.com/)'s documentation.

Reviewer: ```suggestion
All credits go to the Tailwind Labs team for designing this component, found on the [Tailwind CSS website](https://tailwindcss.com/).
```

:::

## Setting up the project

> We are going to use `create-react-app` in thsi demo. For setting up the project, follow the instructions give in the [official Tailwind CSS installation page](https://tailwindcss.com/docs/guides/create-react-app).

Reviewer: ```suggestion
We'll use `create-react-app` for this guide. 
For setting it up, follow the instructions given on the [Tailwind CSS installation page](https://tailwindcss.com/docs/guides/create-react-app).
```

> After you are done with that, add `@mui/base` in the project using:

Reviewer: ```suggestion
Once you're done with the basic setup, install `@mui/base` in the project:
```

```sh
npm install @mui/base
```

## Adding the player markdown

> Next, we are going to add the markdown for the player that already has Tailwind CSS on top.

Reviewer: ```suggestion
Now, create a file called `Player.tsx` and add the markdown below in it, which is already using Tailwind CSS.
```

                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            type="button"
            className="hidden sm:block lg:hidden xl:block"
            aria-label="Next"
          >
            <svg width="24" height="24" fill="none">
              <path
                d="M14 12 6 6v12l8-6Z"
                fill="currentColor"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M18 6v12"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            type="button"
            className="rounded-lg text-xs leading-6 font-semibold px-2 ring-2 ring-inset ring-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500"
          >
            1x
          </button>
        </div>
      </div>
    </div>
  );
});

export default Player;
```

> After this, you should see the player from the picture above. If you try to interact with it, for example use the slider, you will notice that it doesn't do anything. We are going to be adding the functionality in the next steps.

Reviewer: ```suggestion
You should see the player component render as it is on the screenshot above.
However, the component is not yet interactive. 
That's what we'll do in the next step.
```

## Adding an interactive Slider component

> In order to fix the issue with the uninteractive `Slider`, we are going to use the `SliderUnstyled` component from MUI Base.

Reviewer: ```suggestion
Let's start by giving life to the slider. 
For that, we'll use the `SliderUnstyled` component from MUI Base.
```

> Let's add a new file next to the `App.tsx`, called `Slider.tsx`.

Reviewer: ```suggestion
The first step is creating a new file called `Slider.tsx`.
```

> Next we will create a custom `Slider` component that uses the `SliderUnstyled` component.

Reviewer: ```suggestion
Now, copy the code below in the file.
```

```tsx
import * as React from 'react';
import SliderUnstyled, { SliderUnstyledThumbSlotProps, SliderUnstyledProps } from '@mui/base/SliderUnstyled';

const Slider = React.forwardRef(function Slider(
  props: SliderUnstyledProps,
  ref: React.ForwardedRef<HTMLSpanElement>,
){
  return (<SliderUnstyled
    {...props}
    ref={ref}
    componentsProps={{
      thumb: { className: "ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute" }
      root: { className: "w-full relative inline-block h-2 cursor-pointer" },
      rail: { className: "bg-slate-100 dark:bg-slate-700 h-2 w-full rounded-full block absolute" },
      track: { className: "bg-cyan-500 dark:bg-cyan-400 h-2 absolute rounded-full" }
    }}
  />);
});

export default Slider;
```

> You will notice that the `componentsProps` above, defines the class names for the different parts of the slider.
> Most of the classes are copied from the initial markup, with small adjustments in order to make the `Slider` functional.
> Next, let's use this `Slider` in the `Player` component, by replacing the lines 24-29 with:

Reviewer: ```suggestion
To assign specific Tailwind CSS utility classes for each part of the component, we're using `componentsProps`.
Most of them were copied from the original markup with small adjustments now that it is interactive.

Let's add the `Slider` into the `Player` component now:
```

   const { className, ...other } = props;
@@ -21,12 +22,7 @@ const Player = React.forwardRef(function Player(props: { className?: string }, r
         </div>
         <div className="space-y-2">
           <div className="relative">
-            <div className="bg-slate-100 dark:bg-slate-700 rounded-full overflow-hidden">
-              <div className="bg-cyan-500 dark:bg-cyan-400 w-1/2 h-2" role="progressbar" aria-label="music progress" aria-valuenow={1456} aria-valuemin={0} aria-valuemax={4550}></div>
-            </div>
-            <div className="ring-cyan-500 dark:ring-cyan-400 ring-2 absolute left-1/2 top-1/2 w-4 h-4 -mt-2 -ml-2 flex items-center justify-center bg-white rounded-full shadow">
-              <div className="w-1.5 h-1.5 bg-cyan-500 dark:bg-cyan-400 rounded-full ring-1 ring-inset ring-slate-900/5"></div>
-            </div>
+            <Slider step={50} defaultValue={1456} max={4550} min={0} />
           </div>
           <div className="flex justify-between text-sm leading-6 font-medium tabular-nums">
             <div className="text-cyan-500 dark:text-slate-100">24:16</div>
```

> After this you will get the following result:

Reviewer: ```suggestion
You should be seeing this:
```

> <img src="/static/images/base-with-tailwind-css/player-slider.png" alt="Example UI using Tailwind CSS - player" width="638" />

Reviewer: ```suggestion
<img src="/static/images/base-with-tailwind-css/player-slider.png" alt="Screenshot of the media player used as example in the guide, designed by the Tailwind Labs team" width="638" />
```

> You will notice now that the slider is interactive, however it doesn't look exactly as the initial picture.
> This is because we haven't defined the element that represents the dot inside the thumb.
> In order to do this, is not enough to just propagate classes for the thumb, we need to actually render a custom component for it.
> We can do this easily by passing the custom component in the `components` props of the `Slider`.
> This is what we need to do

Reviewer: ```suggestion
Even though the slider is now interactive, it still does not look exactly like the original design.
This is because we haven't defined the element that represents the dot inside the thumb.

To do this, it's not enough to just use classes for the thumb, we need also to render a custom component that gets passed in the `components` prop of the `Slider`.
```

+) {
+  const { ownerState, className, children, ...other } = props;
+  return (<span className={`${className} ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute`} {...other} ref={ref}>
+    <span className="w-1.5 h-1.5 bg-cyan-500 dark:bg-cyan-400 rounded-full ring-1 ring-inset ring-slate-900/5"></span>
+    {children}
+  </span>);
+});
+
 const Slider = React.forwardRef(function Slider(
   props: SliderUnstyledProps,
   ref: React.ForwardedRef<HTMLSpanElement>,
@@ -8,9 +19,11 @@ const Slider = React.forwardRef(function Slider(
   return (<SliderUnstyled
     {...props}
     ref={ref}
+    components={{
+      Thumb,
+    }}
     componentsProps={{
       root: { className: "w-full relative inline-block h-2 cursor-pointer" },
-      thumb: { className: "ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute" },
       rail: { className: "bg-slate-100 dark:bg-slate-700 h-2 w-full rounded-full block absolute" },
       track: { className: "bg-cyan-500 dark:bg-cyan-400 h-2 absolute rounded-full" }
     }}
```

> If you refresh the page now, you will see that the thumb looks idential as in the first picture.

Reviewer: ```suggestion
After refreshing the page, you should see the thumb looking identical to the design now.
```

> All that we did is, created custom component, appended the same classes as we did before on the root element, and before rendering the default children provided by the Slider component, we've added a span that represents the dot inside the thumb.

Reviewer: ```suggestion
What we did was create a custom component, that had the same classes as before now appended to the root element, and before rendering the default children provided by the `Slider`, we added a `<span>` that represents the dot inside the thumb.
```

> This is just a small example, but you can image building much more custom component for each of the slots available in the MUI Base components.

Reviewer: ```suggestion
This is just an example but you can build a lot more custom components for each and every slot available in the MUI Base components.
```

:::warning

> ⚠️ When you are building building custom components for the slots, it is important to always <b>propagate the props sent from the owner component on the root element</b>.

Reviewer: ```suggestion
⚠️ When building custom components for the slots, always <b>propagate the props sent from the owner component on the root element</b>.
```

> This is important, as the props will contain event handlers and aria properties necessaryf or making the component accessible.\

Reviewer: ```suggestion
This is because the props will contain event handlers and aria properties necessary to make the component accessible.
```

> In addition to this, each of the slots receives an `ownerState` object, which contains the props & state of the owner component. This is useful if you want to style the component based on some internal state.

Reviewer: ```suggestion
Additionally, each of the slots receives an `ownerState` object, which contains the props & state of the owner component. 
This is useful if you want to style the component based on some internal state.
```
Might make sense to remove this part from the warning call out, though.

Author: Agree

:::

> ## Adding custom focus visible selector on the buttons

Reviewer: ```suggestion
## Adding a custom `focus-visible` selector on the buttons
```

Author: it's misleading if we use `focus-visible` here as it is valid pseudo selector. We are rather using `focusVisible` state which is in a way polyfill for it.

Author: ```suggestion
## Adding a custom focus visible selector on the buttons
```

> The last step of this guide, will cover how we can add custom styles based on an internal state of the component.

Reviewer: ```suggestion
To finish this guide off, let's see how you can add custom styles based on a component's internal state.
```

> To showcase this, we will create a custom `Button` component, that will use the `focusVisible` state of the `ButtonUnstyled` for applying a cyan ring around the button.

Reviewer: ```suggestion
We'll create a custom `Button` component that uses the `focusVisible` state from the MUI Base `ButtonUnstyled` to apply a cyan ring around it.
```

> This is how the focused button will look like in the end:

Reviewer: ```suggestion
This is how it'll look like:
```

> <img src="/static/images/base-with-tailwind-css/player-buttons.png" alt="Example UI using Tailwind CSS - player" width="638" />

Reviewer: ```suggestion
<img src="/static/images/base-with-tailwind-css/player-buttons.png" alt="Screenshot of a button used as example in the guide, designed by the Tailwind Labs team" width="638" />
```

> Let's add a custom `Button.tsx` file that will have the following content:

Reviewer: ```suggestion
Create a `Button.tsx` file and copy the following code:
```

```tsx
import * as React from 'react';
import ButtonUnstyled, {
  ButtonUnstyledOwnerState,
  ButtonUnstyledProps,
} from '@mui/base/ButtonUnstyled';

const Button = React.forwardRef(function Button(
  props: ButtonUnstyledProps,
  ref: React.ForwardedRef<HTMLButtonElement>,
) {
  const { ...other } = props;
  return (
    <ButtonUnstyled
      componentsProps={{
        root: (state: ButtonUnstyledOwnerState) => ({
          className: `hover:text-cyan-500 transition-colors ${
            state.focusVisible ? 'outline-0 ring-2 ring-cyan-500' : ''
          }`,
        }),
      }}
      {...other}
      ref={ref}
    />
  );
});

export default Button;
```

> You can see how in this demo, we use a `callback` for the `root` inside the `componentsProps` so that we can conditionaly apply classes when the `focusVisible` is true.

Reviewer: ```suggestion
Note that we're using a `callback` for the `root` element inside the `componentsProps`. 
This allows us to conditionally apply utility classes if `focusVisible` is true.
```

> Last, let's replace the all buttons in the intial markup with the new custom `Button` component.

Reviewer: ```suggestion
Now, let's replace all buttons in the initial markup with the new custom `Button` component.
```

             </svg>
-          </button>
-          <button type="button" className="hidden sm:block lg:hidden xl:block" aria-label="Next">
+          </Button>
+          <Button className="hidden sm:block lg:hidden xl:block" aria-label="Next">
             <svg width="24" height="24" fill="none">
               <path d="M14 12 6 6v12l8-6Z" fill="currentColor" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" />
               <path d="M18 6v12" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" />
             </svg>
-          </button>
-          <button type="button" className="rounded-lg text-xs leading-6 font-semibold px-2 ring-2 ring-inset ring-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500">
+          </Button>
+          <Button className="rounded-lg text-xs leading-6 font-semibold px-2 border-2 border-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500 hover:ring-cyan-500">
             1x
-          </button>
+          </Button>
         </div>
       </div>
     </div>
```

> You may noticed that I slightly changed the classes use for the custom borders on some of the buttons so that we can have a consistent focus indicator on all buttons.

Reviewer: ```suggestion
Some classes were slightly changed on some buttons so we have a consistent focus indicator.
```

## Conclusion

> This is what we learned in this guide:

Reviewer: ```suggestion
These are the things we covered in this guide:
```

> ✅ How to apply the Tailwind CSS classes inside the MUI Base components using the `componentsProps` prop, which allows us to target each slot inside the components.\

Reviewer: ```suggestion
✅ How to use Tailwind CSS utility classes to style MUI Base components, using the `componentsProps`prop for targeting specific slots within the component.
```

> ✅ We also learned that in some more involved customization scenarios, we may even need to create custom components for some of the slots. We can then use the `components` prop to pass the custom component. When creating custom components, it is important not to forget to pass all props on the root of the component.\

Reviewer: ```suggestion
✅ How to create custom components for specific slots when in more complex customization scenarios.
We used the `component` prop to pass them into the parent component.
```

The "it is important not to forget..." part might be better suited somewhere else.

> ✅ Lastly, if we want to add some conditional style based on the state of the owner component, we can use a callback as a value inside the `componentsProps` prop.

Reviewer: ```suggestion
✅ How to apply conditional styling based on the owner component's state using a callback as a value for the `componentsProps` prop.
```

> You can find the complete code of this demo under in the following [link](https://github.com/mui/material-ui/tree/master/examples/mui-base-with-tailwind-css).

Reviewer: ```suggestion
Get all the code used in this guide in the [MUI Base with Tailwind CSS example project](https://github.com/mui/material-ui/tree/master/examples/mui-base-with-tailwind-css).
```

```tsx
import * as React from 'react';
import SliderUnstyled, { SliderUnstyledThumbSlotProps, SliderUnstyledProps } from '@mui/base/SliderUnstyled';

const Slider = React.forwardRef(function Slider(
  props: SliderUnstyledProps,
  ref: React.ForwardedRef<HTMLSpanElement>,
){
  return (<SliderUnstyled
    {...props}
    ref={ref}
    componentsProps={{
      thumb: { className: "ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute" }
      root: { className: "w-full relative inline-block h-2 cursor-pointer" },
      rail: { className: "bg-slate-100 dark:bg-slate-700 h-2 w-full rounded-full block absolute" },
      track: { className: "bg-cyan-500 dark:bg-cyan-400 h-2 absolute rounded-full" }
    }}
  />);
});

export default Slider;
```

You will notice that the `componentsProps` above, defines the class names for the different parts of the slider.
Most of the classes are copied from the initial markup, with small adjustments in order to make the `Slider` functional.
Next, let's use this `Slider` in the `Player` component, by replacing the lines 24-29 with:

```diff

> --- a/src/App.tsx
> +++ b/src/App.tsx

Author: ```suggestion
--- a/src/Player.tsx
+++ b/src/Player.tsx
```

<img src="/static/images/base-with-tailwind-css/player-buttons.png" alt="Example UI using Tailwind CSS - player" width="638" />

Let's add a custom `Button.tsx` file that will have the following content:

```tsx
import * as React from 'react';
import ButtonUnstyled, {
  ButtonUnstyledOwnerState,
  ButtonUnstyledProps,
} from '@mui/base/ButtonUnstyled';

const Button = React.forwardRef(function Button(
  props: ButtonUnstyledProps,
  ref: React.ForwardedRef<HTMLButtonElement>,
) {
  const { ...other } = props;
  return (
    <ButtonUnstyled
      componentsProps={{
        root: (state: ButtonUnstyledOwnerState) => ({
          className: `hover:text-cyan-500 transition-colors ${
            state.focusVisible ? 'outline-0 ring-2 ring-cyan-500' : ''
          }`,
        }),
      }}
      {...other}
      ref={ref}
    />
  );
});

export default Button;
```

You can see how in this demo, we use a `callback` for the `root` inside the `componentsProps` so that we can conditionaly apply classes when the `focusVisible` is true.
Last, let's replace the all buttons in the intial markup with the new custom `Button` component.

```diff

> --- a/src/App.tsx
> +++ b/src/App.tsx

Author: ```suggestion
--- a/src/Player.tsx
+++ b/src/Player.tsx
```

# Interoperability with Tailwind CSS

<p class="description">Learn how to use Tailwind CSS with MUI Base components.</p>

## Getting started

The goal of this guide is to teach you how to style MUI Base components using Tailwind CSS.
If you don't have experience with Tailwind CSS, we recommend you study that first before diving into this guide.
We're not going to cover all the steps to build the UI we'll use as an example but rather how to insert MUI Base components, making it interactive and accessible, and style it with Tailwind CSS.

This guide result is an interactive media player.
That's how it's going to look like in the end.

<img src="/static/images/base-with-tailwind-css/player-final.png" alt="Screenshot of the media player used as example in the guide, designed by the Tailwind Labs team" width="638" />

:::info
All credits go to the Tailwind Labs team for designing this component, found on the [Tailwind CSS website](https://tailwindcss.com/).
:::

## Setting up the project

> We'll use `create-react-app` for this guide.

Author: Restructured a bit.

> For setting it up, follow the instructions given on the [Tailwind CSS installation page](https://tailwindcss.com/docs/guides/create-react-app).

Author: I would like to not have to maintain how Tailwind CSS is configured with create-react-app. I would prefer linking to their site, expecting that they would update the docs if anything changes.

Once you're done with the basic setup, install `@mui/base` in the project:

```sh
npm install @mui/base
```

> ## Adding the player markdown

Author: I don't know why I used markdown everywhere 😅

                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            type="button"
            className="hidden sm:block lg:hidden xl:block"
            aria-label="Next"
          >
            <svg width="24" height="24" fill="none">
              <path
                d="M14 12 6 6v12l8-6Z"
                fill="currentColor"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M18 6v12"
                stroke="currentColor"
                stroke-width="2"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <button
            type="button"
            className="rounded-lg text-xs leading-6 font-semibold px-2 ring-2 ring-inset ring-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500"
          >
            1x
          </button>
        </div>
      </div>
    </div>
  );
});

export default Player;
```

> Next, add the `Player` component in the `App.tsx` file.

Author: Done 👍

+  return (<span className={`${className} ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute`} {...other} ref={ref}>
+    <span className="w-1.5 h-1.5 bg-cyan-500 dark:bg-cyan-400 rounded-full ring-1 ring-inset ring-slate-900/5"></span>
+    {children}
+  </span>);
+});
+
 const Slider = React.forwardRef(function Slider(
   props: SliderUnstyledProps,
   ref: React.ForwardedRef<HTMLSpanElement>,
@@ -8,9 +19,11 @@ const Slider = React.forwardRef(function Slider(
   return (<SliderUnstyled
     {...props}
     ref={ref}
+    components={{
+      Thumb,
+    }}
     componentsProps={{
       root: { className: "w-full relative inline-block h-2 cursor-pointer" },
-      thumb: { className: "ring-cyan-500 dark:ring-cyan-400 ring-2 w-4 h-4 -mt-1 -ml-2 flex items-center justify-center bg-white rounded-full shadow absolute" },
       rail: { className: "bg-slate-100 dark:bg-slate-700 h-2 w-full rounded-full block absolute" },
       track: { className: "bg-cyan-500 dark:bg-cyan-400 h-2 absolute rounded-full" }
     }}
```

After refreshing the page, you should see the thumb looking identical to the design now.

> What we did was create a custom component, that had the same classes as before now appended to the root element, and before rendering the default children provided by the `Slider`, we added a `<span>` that represents the dot inside the thumb.

Author: I've tried my best to explain it better :)

-          </button>
-          <button type="button" className="hidden sm:block lg:hidden xl:block" aria-label="Next">
+          </Button>
+          <Button className="hidden sm:block lg:hidden xl:block" aria-label="Next">
             <svg width="24" height="24" fill="none">
               <path d="M14 12 6 6v12l8-6Z" fill="currentColor" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" />
               <path d="M18 6v12" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" />
             </svg>
-          </button>
-          <button type="button" className="rounded-lg text-xs leading-6 font-semibold px-2 ring-2 ring-inset ring-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500">
+          </Button>
+          <Button className="rounded-lg text-xs leading-6 font-semibold px-2 border-2 border-slate-500 text-slate-500 dark:text-slate-100 dark:ring-0 dark:bg-slate-500 hover:ring-cyan-500">
             1x
-          </button>
+          </Button>
         </div>
       </div>
     </div>
```

Some classes were slightly changed on some buttons so we have a consistent focus indicator.

> ## Conclusion

Reviewer 2: This is a silly pet peeve of mine 😛 but I don't like using the word "Conclusion" as the header for the conclusion of an article.

```suggestion
## What we learned
```

These are the things we covered in this guide:

> ✅ How to use Tailwind CSS utility classes to style MUI Base components, using the `componentsProps`prop for targeting specific slots within the component.\

Reviewer 2: ```suggestion
✅ How to use Tailwind CSS utility classes to style MUI Base components, using the `componentsProps` prop for targeting specific slots within the component.\
```

> ✅ How to use Tailwind CSS utility classes to style MUI Base components, using the `componentsProps`prop for targeting specific slots within the component.
> ✅ How to create custom components for specific slots in more complex customization scenarios.
> We used the `component` prop to pass them into the parent component.
> ✅ How to apply conditional styling based on the owner component's state using a callback as value for the `componentsProps` prop.

Author: ```suggestion
✅ How to use Tailwind CSS utility classes to style MUI Base components, using the `componentsProps`prop for targeting specific slots within the component.\
✅ How to create custom components for specific slots in more complex customization scenarios.
We used the `component` prop to pass them into the parent component.\
✅ How to apply conditional styling based on the owner component's state using a callback as value for the `componentsProps` prop.
```
